### PR TITLE
Create generic API for crypto protocol data

### PR DESCRIPTION
**Pull request by djc**
As discussed in #638. @kim, what do you think? @DemiMarie-parity, you might also be interested.

Is `AuthenticationData` still the best name? It might as well be called `Data` in `crypto::Session`, could also be `CryptoData` to be even more generic.

### CODE REVIEW

**Comment by kim** [comment]
Looks great!

I think `Data` works, as it’s pronounced “session data”.

**Comment by djc** [comment]
I figured I'd actually try to make quinn generic, and it turned out to be fairly straightforward, mostly adding lots of generic parameters and adding some extra bounds.

**Comment by DemiMarie-parity** [comment]
Would it help if I wrote tests for this feature?

**Comment by djc** [comment]
Yes, I would love if you could add some tests!

File: quinn-proto/src/endpoint.rs [hunk 1]
```diff
@@ -848,7 +848,7 @@ enum ConnectionOpts<S: crypto::Session> {
 /// Errors in the parameters being used to create a new connection
 ///
 /// These arise before any I/O has been performed.
-#[derive(Debug, Error, Clone, PartialEq, Eq)]
+#[derive(Debug, Error, Clone, PartialEq)]
```
[the remark this answers is not in the document]
Comment (djc) [reply quinn-proto/src/endpoint.rs]: Oh, I forgot to revert that after experimenting with returning error enums.

[the remark this answers is not in the document]
Comment (djc) [reply quinn-proto/src/endpoint.rs]: Fixed.

File: quinn-proto/src/shared.rs [hunk 2]
```diff
@@ -562,7 +592,7 @@ where
 }
 
 /// Errors in the configuration of an endpoint
-#[derive(Debug, Error, Clone, PartialEq, Eq)]
+#[derive(Debug, Error, Clone, PartialEq)]
```
[the remark this answers is not in the document]
Comment (djc) [reply quinn-proto/src/shared.rs]: Fixed.

File: quinn-proto/src/crypto.rs [hunk 3]
```diff
@@ -23,24 +23,29 @@ use crate::{
...
     /// Type used to sign various values
     type HmacKey: HmacKey;
     /// Type used to represent packet protection keys
-    type Keys: Keys + Sized;
+    type Keys: Keys;
     /// Type used to hold configuration for server sessions
     type ServerConfig: ServerConfig<Self>;
 
-    /// Get the negotiated ALPN protocol
+    /// Get the data agreed upon during the cryptographic handshake
```
[the remark this answers is not in the document]
Comment (djc) [reply quinn-proto/src/crypto.rs]: Expanded.

File: quinn-proto/src/crypto.rs [hunk 4]
```diff
@@ -86,7 +86,7 @@ pub trait Session: Sized {
 }
 
 /// Client-side configuration for the crypto protocol
-pub trait ClientConfig<S>
+pub trait ClientConfig<S>: Clone + Default
```
[the remark this answers is not in the document]
Comment (djc) [reply quinn-proto/src/crypto.rs]: I've actually removed the `Default` boundary. At least in the case of `Arc<rustls::ClientConfig>`, coherence prevents us from implementing `Default` ourselves. We could/should potentially wrap it in a newtype, but I'd want to defer that from this PR.

File: quinn-proto/src/crypto.rs [hunk 5]
```diff
@@ -24,6 +24,9 @@ pub(crate) mod ring;
 /// TLS interface based on rustls
 #[cfg(feature = "rustls")]
 pub(crate) mod rustls;
+/// Public interface TLS types
+#[cfg(feature = "rustls")]
+pub mod types;
```
[the remark this answers is not in the document]
Comment (djc) [reply quinn-proto/src/crypto.rs]: Fair enough.

File: quinn/Cargo.toml [hunk 6]
```diff
@@ -15,9 +15,10 @@ edition = "2018"
 all-features = true
 
 [features]
-default = ["native-certs", "ct-logs"]
 # Trust the contents of the OS certificate store by default
-native-certs = [ "rustls-native-certs" ]
+default = ["native-certs", "certificate-transparency"]
+certificate-transparency = ["proto/certificate-transparency"]
```
[the remark this answers is not in the document]
Comment (djc) [reply quinn/Cargo.toml]: Added.

File: quinn-proto/Cargo.toml [hunk 7]
```diff
@@ -18,14 +18,18 @@ maintenance = { status = "experimental" }
 
 [features]
 default = ["tls-rustls"]
+certificate-transparency = ["ct-logs"]
 tls-rustls = ["rustls", "webpki", "ring"]
+native-certs = ["rustls-native-certs"]
```
[the remark this answers is not in the document]
Comment (djc) [reply quinn-proto/Cargo.toml]: Added.

File: quinn/src/connection.rs [hunk 8]
```diff
@@ -299,6 +299,18 @@ impl Connection {
         self.0.lock().unwrap().inner.remote_address()
     }
 
+    /// The connection's authentication data
```
[the remark this answers is not in the document]
Comment (djc) [reply quinn/src/connection.rs]: I instead reused your "Data conveyed by the peer during the handshake, including cryptographic identity" from the other suggestion here.

File: quinn/Cargo.toml [hunk 9]
```diff
@@ -15,9 +15,10 @@ edition = "2018"
 all-features = true
 
 [features]
-default = ["native-certs", "ct-logs"]
 # Trust the contents of the OS certificate store by default
```
[the remark this answers is not in the document]
Comment (djc) [reply quinn/Cargo.toml]: Yup, done.

File: quinn-proto/src/crypto.rs [hunk 10]
```diff
@@ -30,6 +30,8 @@ pub mod types;
 
 /// A cryptographic session (commonly TLS)
 pub trait Session: Sized {
+    /// Authentication data to identify the peer's identity
```
[the remark this answers is not in the document]
Comment (djc) [reply quinn-proto/src/crypto.rs]: Done.

**Comment by djc** [comment]
Thanks for the review, hope it wasn't too onerous. I did an initial attempt at introducing a rustls feature flag to quinn, but I have not yet figured out how to handle the examples (which all require rustls). I tried to replicate tokio's setup for this without any luck.

File: quinn-proto/Cargo.toml [hunk 11]
```diff
@@ -18,14 +18,20 @@ maintenance = { status = "experimental" }
 
 [features]
 default = ["tls-rustls"]
+# Enable checking of certificate transparency logs
```
Comment (Ralith) [inline quinn-proto/Cargo.toml]: Checking of certificate transparency logs is an always-available rustls feature; rather, this bakes in in a specific set of logs and configures them by default, similar to how we formerly offered webpki-roots for baked-in root certs.

File: quinn/Cargo.toml [hunk 12]
```diff
@@ -15,9 +15,11 @@ edition = "2018"
 all-features = true
 
 [features]
-default = ["native-certs", "ct-logs"]
+default = ["native-certs", "certificate-transparency"]
+# Enable checking of certificate transparency logs
```
Comment (Ralith) [inline quinn/Cargo.toml]: As above

**Comment by Ralith** [comment]
> but I have not yet figured out how to handle the examples (which all require rustls). I tried to replicate tokio's setup for this without any luck.

Enumerating the examples explicitly on Cargo.toml with a suitable `required-features` clause should do the trick, I'd think.

File: quinn-proto/Cargo.toml (hunk 11 above)
Comment (djc) [reply quinn-proto/Cargo.toml]: Replaced with "Use Google's list of CT logs to enable certificate transparency checks".

File: quinn/Cargo.toml (hunk 12 above)
Comment (djc) [reply quinn/Cargo.toml]: As above.

**Comment by djc** [comment]
Alright, I've added a commit that makes rustls/webpki optional for quinn.

**Comment by kim** [comment]
Great work!

**Pull request merged**